inc/ctest.h

```diff
@@ -204,55 +204,68 @@ typedef void* void_ptr;
-#endif
 
 extern C_LINKAGE char* ctest_sprintf_char(const char* format, ...);
 extern C_LINKAGE void ctest_sprintf_free(char* string);
 
+#define CTEST_DECLARE_EQUALITY_ASSERTION_FUNCTIONS_FOR_TYPE(type) \
+extern C_LINKAGE void MU_C2(type,_AssertAreEqual)(type left, type right, char* ctest_message); \
+extern C_LINKAGE void MU_C2(type,_AssertAreNotEqual)(type left, type right, char* ctest_message);
+
+#define CTEST_ASSERT_ARE_EQUAL_IMPL_FOR_TYPE(type) \
+    char expectedString[1024]; \
+    char actualString[1024]; \
+    MU_C2(type,_ToString)(expectedString, sizeof(expectedString), left); \
+    MU_C2(type,_ToString)(actualString, sizeof(actualString), right); \
+    if (MU_C2(type,_Compare)(left, right)) \
+    { \
+        LogError("  Assert failed in line %d %s Expected: %s, Actual: %s\n", __LINE__, (ctest_message == NULL) ? "" : ctest_message, expectedString, actualString); \
+        ctest_sprintf_free(ctest_message); \
+        if (g_CurrentTestFunction != NULL) *g_CurrentTestFunction->TestResult = TEST_FAILED; \
+        do_jump(&g_ExceptionJump, expectedString, actualString); \
+    } \
+    ctest_sprintf_free(ctest_message);
+
+#define CTEST_ASSERT_ARE_NOT_EQUAL_IMPL_FOR_TYPE(type) \
+    char expectedString[1024]; \
+    char actualString[1024]; \
+    MU_C2(type,_ToString)(expectedString, sizeof(expectedString), left); \
+    MU_C2(type,_ToString)(actualString, sizeof(actualString), right); \
+    if (!MU_C2(type,_Compare)(left, right)) \
+    { \
+        LogError("  Assert failed in line %d %s Expected: %s, Actual: %s\n", __LINE__, (ctest_message == NULL) ? "" : ctest_message, expectedString, actualString); \
+        ctest_sprintf_free(ctest_message); \
+        if (g_CurrentTestFunction != NULL) *g_CurrentTestFunction->TestResult = TEST_FAILED; \
+        do_jump(&g_ExceptionJump, expectedString, actualString); \
+    } \
+    ctest_sprintf_free(ctest_message);
+
+#define CTEST_DEFINE_STATIC_EQUALITY_ASSERTION_FUNCTIONS_FOR_TYPE(type) \
+static void MU_C2(type,_AssertAreEqual)(type left, type right, char* ctest_message) \
+{ \
+    CTEST_ASSERT_ARE_EQUAL_IMPL_FOR_TYPE(type) \
+} \
+static void MU_C2(type,_AssertAreNotEqual)(type left, type right, char* ctest_message) \
+{ \
+    CTEST_ASSERT_ARE_NOT_EQUAL_IMPL_FOR_TYPE(type) \
+}
+
+#define CTEST_DEFINE_EXTERN_EQUALITY_ASSERTION_FUNCTIONS_FOR_TYPE(type) \
+void MU_C2(type,_AssertAreEqual)(type left, type right, char* ctest_message) \
+{ \
+    CTEST_ASSERT_ARE_EQUAL_IMPL_FOR_TYPE(type) \
+} \
+void MU_C2(type,_AssertAreNotEqual)(type left, type right, char* ctest_message) \
+{ \
+    CTEST_ASSERT_ARE_NOT_EQUAL_IMPL_FOR_TYPE(type) \
+}
+
+
 #define CTEST_COMPARE(toStringType, cType) \
     typedef cType toStringType; \
+    static int MU_C2(toStringType,_Compare)(toStringType left, toStringType right); \
+    static void MU_C2(toStringType,_ToString)(char* string, size_t bufferSize, cType value); \
+    CTEST_DEFINE_STATIC_EQUALITY_ASSERTION_FUNCTIONS_FOR_TYPE(toStringType) \
```

Comment:
Can you please add a comment here on the reason why we are shoehorning the asserts here? It is a valid reason to shoehorn, just think it's worth documenting so that future readers understand why we did that.